Django 2 Compatible Changes

tvle236 opened the pull request:
Made the necessary changes to ensure the code is Django 2 compatible:

- Made some changes to ensure the code works with Django 1.11 and Django 2.2, as well as updating the test cases to run against both Django 1.11 and Django 2.2
- Updated the README, requirements and other parts of the code to make sure we reference Django 1.11 and higher.
- Made changes to the test cases to ensure we only load the required middleware to test functionality. This should help reduce interference from other middleware.
- Minor code clean up

Refs: [PAS-197](https://sdelements.atlassian.net/browse/PAS-197)

karally commented on README.md:
```diff
@@ -12,10 +12,10 @@ Automatically generated documentation of `django-security` is available on Read
+* Django >= 1.11
 
-For Django < 1.8 use django-security==0.9.4.
+For Django < 1.11 use django-security==0.9.5. For Django < 1.8 use django-security==0.9.4.
```
Out of curiosity, why is it 0.9.5 for Django < 1.11? Does our current version support Django 1.8+?

Also strange thing I don't see 0.9.4 on PyPI.

tvle236 replied:
It actually should be `0.11.3` for Django < 1.11, as that's the last version that will support some of the older version. Fixed to it.

That's weird about 0.9.4. Let me take a look into it.

karally commented on setup.py:
```diff
@@ -35,14 +35,10 @@ def run(self):
-          'Framework :: Django :: 1.9',
-          'Framework :: Django :: 1.10',
           'Framework :: Django :: 1.11',
+          'Framework :: Django :: 2.22',
```
Is that an extra 2?

tvle236 replied:
Yeah there's an extra `2`, dropped it.

RouganStriker commented on setup.py:
```diff
@@ -35,14 +35,10 @@ def run(self):
-          'Framework :: Django :: 1.9',
-          'Framework :: Django :: 1.10',
           'Framework :: Django :: 1.11',
+          'Framework :: Django :: 2.2',
```
Shall we add 3.0 too since the tests are passing?

tvle236 replied:
Sure, added it in.

RouganStriker commented:
Code looks good to me. I can do a bit of dev qaing unless Kara is already doing it :eyes:

karally commented:
Do we still need this import bit? `MiddlewareMixin` is introduced in 1.10 and we only support 1.11+ now.

(I can't seem to comment on code outside of PR changes. This is in `middleware.py`)

```
try:
    from django.utils.deprecation import MiddlewareMixin
except ImportError:
    MiddlewareMixin = object
```

tvle236 commented:
> Do we still need this import bit? `MiddlewareMixin` is introduced in 1.10 and we only support 1.11+ now.
> 
> (I can't seem to comment on code outside of PR changes. This is in `middleware.py`)
> 
> ```
> try:
>     from django.utils.deprecation import MiddlewareMixin
> except ImportError:
>     MiddlewareMixin = object
> ```

Yeah we no longer need it. Removed it.

karally commented:
+1

DevQA functionality affected per middleware:

- `SessionExpiryPolicyMiddleware`
  - Session expiry and inactivity timeout work as expected

- `LoginRequiredMiddleware`
  - Unauthenticated requests redirects to the login page and URL is recorded in the `next` param
  - Deactivated user while they are logged in redirects them to login page

- `MandatoryPasswordChangeMiddleware`
  - Redirect to password change page for users that have an expired PasswordExpiry object associated when:
    - Attempt to login
    - Previously logged in and attempts to access a page

- `NoConfidentialCachingMiddleware`
  - No cache-control response headers on whitelisted URLs. They are present on other requests.

- `StrictTransportSecurityMiddleware`
  - `Strict-Transport-Security` response header gets added for login and logouts

- `ClearSiteDataMiddleware`
  - `clear-site-data` response header gets added for logout, and not on other requests


@RouganStriker  I'm not sure if I covered all possible cases, so if you have more ideas please add more to the list 🙏

RouganStriker commented:
StrictTransportSecurityMiddleware, ContentNoSniff, XssProtectMiddleware have equivalents in the core django library: https://docs.djangoproject.com/en/2.2/ref/settings/; might be candidate for removal.

XFrameOptionsMiddleware is a semi-equivalent in https://docs.djangoproject.com/en/2.2/ref/clickjacking/ but it doesn't seem to provide the easy list of exempt URLs.

P3PPolicyMiddleware is to set P3P which is a very old standard that currently is not supported by any modern browser (As of windows 10, internet explorer and edge no longer support P3P). Might be okay to remove?

tvle236 commented:
If we are going to drop them, we should add a deprecation warning for each of those middlewares to give enough time for developers to find an alternative approach. I have added a deprecation message to all 5 of those specified.

The only one I might consider keeping as it is significantly different enough is `XFrameOptionsMiddleware` as they may serve different purposes.

RouganStriker commented on security/middleware.py:
```diff
@@ -214,6 +219,14 @@ class XssProtectMiddleware(BaseMiddleware):
+        super().__init__(get_response)
+        warnings.warn('The middleware "{name}" will no longer be supported in '
+        'future releases of this library. Refer to {url} for an '
+        'alternative approach.'.format(
```
Could we provide a little more detail such as mentioning the django settings that replace the middleware? Can keep the link to the docs as an additional info link.

RouganStriker commented:
ContentSecurityPolicyMiddleware (Forgot to include this one in list of middlewares used by app
* content-security-policy header is included in API responses with middleware include; header is not included without middleware
* ShimmedSAMLContentSecurityPolicyMiddleware (SDE customization) includes SAML IdP in CSP header value (Bug: Requires server restart to populate)

**Rest of the middlewares:**

DoNotTrackMiddleware
* Sending request with header DNT=1 causes response to return back DNT=1 in repsonse headers as well

XssProtectMiddleware
* x-xss-protection header is included in responses

ContentNoSniff
* x-content-options header with value nosniff is included in responses

XFrameOptionsMiddleware
* x-frame-options header is included in responses

P3PPolicyMiddleware
* With P3P_COMPACT_POLICY declared in settings file, p3p header is included in responses

Middlewares are working as expected, +1 after deprecation warnings are added.

RouganStriker commented:
> If we are going to drop them, we should add a deprecation warning for each of those middlewares to give enough time for developers to find an alternative approach. I have added a deprecation message to all 5 of those specified.
> 
> The only one I might consider keeping as it is significantly different enough is `XFrameOptionsMiddleware` as they may serve different purposes.

Yea that was the only one I was on the fence about. Perhaps we can still mention that Django has a builtin alternative that might suit some use cases?

tvle236 commented:
Updated the deprecation messages to be more helpful, providing guidance on the settings of interest. Also changed the deprecation of `XFrameOptionsMiddleware` to provide a reference to Django's implementation to help guide developers to more official implementations that may fit their use-case.

RouganStriker commented:
One last thing, could we add the deprecation warnings to the README as well?

karally commented:
Another +1 post docs update:

Tested that the deprecation warnings successfully get triggered on app load:

```
/Users/kwang/Revisions/django-security/security/middleware.py:951: UserWarning: The middleware "P3PPolicyMiddleware" will no longer be supported in future releases of this library.
  'future releases of this library.'.format(name=self.__class__.__name__))
/Users/kwang/Revisions/django-security/security/middleware.py:528: UserWarning: An official middleware "XFrameOptionsMiddleware" is supported by Django. Refer to https://docs.djangoproject.com/en/1.11/ref/clickjacking/ to see if its approach fits the use case.
  url=DJANGO_CLICKJACKING_MIDDLEWARE_URL))
/Users/kwang/Revisions/django-security/security/middleware.py:229: UserWarning: The middleware "XssProtectMiddleware" will no longer be supported in future releases of this library. Refer to https://docs.djangoproject.com/en/1.11/ref/middleware/#django.middleware.security.SecurityMiddleware for an alternative approach with regards to the settings: SECURE_BROWSER_XSS_FILTER
  settings="SECURE_BROWSER_XSS_FILTER"))
/Users/kwang/Revisions/django-security/security/middleware.py:337: UserWarning: The middleware "ContentNoSniff" will no longer be supported in future releases of this library. Refer to https://docs.djangoproject.com/en/1.11/ref/middleware/#django.middleware.security.SecurityMiddleware for an alternative approach with regards to the settings: SECURE_CONTENT_TYPE_NOSNIFF
  settings="SECURE_CONTENT_TYPE_NOSNIFF"))
/Users/kwang/Revisions/django-security/security/middleware.py:892: UserWarning: The middleware "StrictTransportSecurityMiddleware" will no longer be supported in future releases of this library. Refer to https://docs.djangoproject.com/en/1.11/ref/middleware/#django.middleware.security.SecurityMiddleware for an alternative approach with regards to the settings: SECURE_HSTS_SECONDS, SECURE_HSTS_INCLUDE_SUBDOMAINS, SECURE_HSTS_PRELOAD
  "SECURE_HSTS_PRELOAD",
```

RouganStriker commented:
+1 doc changes look good. Let's merge and then tag a new release
